sdks/go/pkg/beam/io/synthetic/source.go

```diff
@@ -0,0 +1,151 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package synthetic contains transforms for creating synthetic pipelines.
+// Synthetic pipelines are pipelines that simulate the behavior of possible
+// pipelines in order to test performance, splitting, liquid sharding, and
+// various other infrastructure used for running pipelines. This category of
+// tests are not concerned with the correctness of the elements themselves, but
+// need to simulate transforms that output many elements throughout varying
+// pipeline shapes.
+package synthetic
+
+import (
+	"github.com/apache/beam/sdks/go/pkg/beam"
+	"github.com/apache/beam/sdks/go/pkg/beam/io/rtrackers/offsetrange"
+	"math/rand"
+	"time"
+)
+
+// Source creates a synthetic source transform that emits randomly
+// generated KV<[]byte, []byte> elements.
+//
+// This transform accepts a PCollection of SourceConfig, where each SourceConfig
+// determines the synthetic source's behavior for that element.
```

Comment:
You may want to consider a small pipeline example snippet that feeds into this using beam.Create and DefaultSourceConfig() or similar, if only to hint folks towards DefaultSourceConfig.

WRT Builders, I agree that a Builder api for SourceConfigs would be valuable in this case, even if they just return new instances of a SourceConfig. It would be more concise than requiring a new line per field change.

`synthetic.DefaultSourceConfig().SetNumElements(###).SetInitialSplits(###)`
vs the "natural" Go inclination of:
`synthetic.SourceConfig{NumElements: ###, InitialSplits: ###}`

which is a dozen or so characters shorter.
The only downside of the latter is introducing new fields that *must* have a nonzero default to work properly. Do any of the configurations in Java and Python have that property?

Reply:
Also, added the example.

Reply:
output_records_per_input_record is problematic, but doesn't apply to SourceConfig I would think... (SrcCfg?) That seems like a perDoFn configuration.

Unless that's the "NumElements" you were mentioning?

Reply:
No, it applies to the synthetic steps, so it would be in an upcoming StepConfig, but that would probably still have a similar interface.

Reply:
I think I'm coming to the position that it might be overengineering to protect users, since the risky fields are known with the first version. 

It's probably better to simply panic/error out with a clear message at some early stage if it's configured incorrectly, if graceful behavior isn't possible, and it's not typically the value that needs to be set (eg. NumElements or OutputsPerInput).

Regardless, this is a good exercise to consider, whatever you choose to do.